### ISSUE TITLE
Launching with a main() function?

### ISSUE THREAD

**Issue by thomasnield** [edvin/tornadofx#6]
I might sound like a complete amateur here. How exactly do I launch a TornadoFX App class with a main method? Whenever I try to configure IDEA to launch my `MainApp` class, it does not find my `main()` function. 

```kotlin
class MainApp : App() {
    override val primaryView = HelloWorld::class

    init {
        //importStylesheet("/styles.css")
    }

    class HelloWorld : View() {
        override val root = HBox(Label("Hello world"))
    }

    //IDEA does not see this main() function
    fun main(args: Array<String>) {
        launch("")
    }
}
```

**Comment by edvin** [comment]
I'm assuming you get this example to work in IDEA when you remove the main method, right? Your problem is that you want to override the startup sequence somehow?

First of all, your main method is not static, it's just a method sitting in a class, so Java doesn't recognize it as an application entry point.

Normally you declare Kotlin main methods as a package level construct, ie. outside your App class:

```kotlin
fun main(args: Array<String>) {
    Application.launch(MainApp::class.java, *args)
}
```

You would then tell IDEA that your main class is called `MainAppKt`.

But why do you need to provide the main method? Doesn't IntelliJ recognise `MainApp` as runnable? Without a main method you should simply specify `MainApp` and it should work.

**Comment by edvin** [comment]
@thomasnield see my edit above - maybe you chose the `Kotlin` run configuration instead of the one called `Application`?

**Comment by edvin** [comment]
I created a page in the Wiki where I mention this IDEA specific caveat :)

[[https://github.com/edvin/tornadofx/wiki/Application-Startup]]

**Comment by thomasnield** [comment]
Perfect! Yeah I was just trying to get it to run. Creating the IDEA Application configuration did the trick. Thanks a lot and I look forward to playing with this library.

**Comment by edvin** [comment]
Nice :) Let me know if you need any further help :)

**Comment by thomasnield** [comment]
Thank you sir, will do!

**Comment by hastebrot** [comment]
Yeah, `main()` needs to be on file level (i.e. `FooKt` for the file `foo.kt`). Luckily Kotlin allows multiple `main`s within a package, so it allows multiple `Application`s.

By the way: IDEA treats JavaFX `Application`s in Java to allow to omit the `main` method. I opened a feature request, to have the same for Kotlin: https://youtrack.jetbrains.com/issue/KT-8773

**Comment by edvin** [comment]
Thanks, @hastebrot - that's what I noticed as well. I voted for KT-8773 as well :) Good to get the ball rolling on that.

**Comment by thomasnield** [comment]
@hastebrot  I just started to figure that out regarding the file level as well as the omission of the `main()` method. Thanks for filing that as it would be convenient.

**Comment by thomasnield** [comment]
Me as well.

**Comment by thomasnield** [comment]
I hate to tangent, but did I build this 3-tabbed pane optimally? ? 

```
class MainApp : App() {
    override val primaryView = MainView::class

    class MainView : View() {

        override val root = with(TabPane()) {
            with (tabs) {
                add(Tab("Tab 1"))
                add(Tab("Tab 2"))
                add(Tab("Tab 3"))
                tabs
            }
            this
        }
    }
}
```

**Comment by hastebrot** [comment]
At least the second element of the `HBox` could be created like this:

~~~kotlin
class Main : View() {
    override val root = StackPane()

    init {
        with(root) {
            setPrefSize(500.0, 500.0)
            hbox {
               vbox {
                   button("Bttn 1")
                   button("Bttn 2")
               }
            }
        }
    }
}

**Comment by hastebrot** [comment]
I wonder why IntelliJ doesn't suggest me `tabpane` and `tab` from `tornadofx/Builders.kt`.

**Comment by hastebrot** [comment]
Or do you want to create the form without Builders?

**Comment by thomasnield** [comment]
I was wondering that too. Is there a function for tabpane?

**Comment by hastebrot** [comment]
Yes, it is right at the top of `tornadofx/Builders.kt`.

**Comment by thomasnield** [comment]
I must have overlooked that. I assumed the builder was missing but that does not appear to be the case. Odd. That aside I'm really liking this library. This comes a long way from my messy Swing applications from back in the day. 

I'm going to use [RxJavaFX](https://github.com/ReactiveX/RxJavaFX) with this next to get some reactive stuff going. Please let me know if we need to file a bug with JetBrain on the `tabpane` issue.

```
class MainApp : App() {
    override val primaryView = MainView::class

    class MainView : View() {

        override val root = StackPane()

        init {
            with(root) {
                setPrefSize(500.0, 500.0)
                hbox {
                    this += with(TabPane()) {
                        with (tabs) {
                            add(Tab("Tab 1"))
                            add(Tab("Tab 2"))
                            add(Tab("Tab 3"))
                        }
                        this
                    }
                    vbox {
                        button("Bttn 1")
                        button("Bttn 2")
                    }
                }
            }
        }
    }
}
```

**Comment by hastebrot** [comment]
>a bug with JetBrain on the tabpane issue

It's all ok. IntelliJ didn't find it, because it doesn't exist in TornadoFX 1.2.2.

**Comment by thomasnield** [comment]
I was just about to post that because my decompiled `BuildersKt.class` didn't have it. 

```
// IntelliJ API Decompiler stub source generated from a class file
// Implementation of methods is not available

package tornadofx

private fun <T : javafx.scene.Node> opcr(pane: javafx.scene.layout.Pane, node: T, op: (T.() -> kotlin.Unit)? /* = compiled code */): T { /* compiled code */ }

public fun javafx.scene.layout.Pane.button(text: kotlin.String /* = compiled code */, op: (javafx.scene.control.Button.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.Button { /* compiled code */ }

public fun javafx.scene.layout.Pane.checkbox(text: kotlin.String? /* = compiled code */, op: (javafx.scene.control.CheckBox.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.CheckBox { /* compiled code */ }

public fun javafx.scene.layout.Pane.datepicker(op: (javafx.scene.control.DatePicker.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.DatePicker { /* compiled code */ }

public fun javafx.scene.layout.Pane.gridpane(op: (javafx.scene.layout.GridPane.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.layout.GridPane { /* compiled code */ }

public fun javafx.scene.layout.Pane.hbox(spacing: kotlin.Double? /* = compiled code */, children: kotlin.collections.Iterable<javafx.scene.Node>? /* = compiled code */, op: (javafx.scene.layout.HBox.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.layout.HBox { /* compiled code */ }

public fun javafx.scene.layout.Pane.imageview(url: kotlin.String? /* = compiled code */, op: (javafx.scene.image.ImageView.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.image.ImageView { /* compiled code */ }

public fun javafx.scene.layout.Pane.label(text: kotlin.String /* = compiled code */, op: (javafx.scene.control.Label.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.Label { /* compiled code */ }

public fun javafx.scene.layout.Pane.progressBar(initialValue: kotlin.Double? /* = compiled code */, op: (javafx.scene.control.ProgressBar.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.ProgressBar { /* compiled code */ }

public fun javafx.scene.layout.Pane.progressIndicator(op: (javafx.scene.control.ProgressIndicator.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.ProgressIndicator { /* compiled code */ }

public fun javafx.scene.layout.HBox.spacer(prio: javafx.scene.layout.Priority /* = compiled code */, op: (javafx.scene.layout.Pane.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.layout.Pane { /* compiled code */ }

public fun javafx.scene.layout.VBox.spacer(prio: javafx.scene.layout.Priority /* = compiled code */, op: (javafx.scene.layout.Pane.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.layout.Pane { /* compiled code */ }

public fun javafx.scene.layout.Pane.stackpane(initialChildren: kotlin.collections.Iterable<javafx.scene.Node>? /* = compiled code */, op: (javafx.scene.layout.StackPane.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.layout.StackPane { /* compiled code */ }

public fun javafx.scene.layout.Pane.textarea(value: kotlin.String? /* = compiled code */, op: (javafx.scene.control.TextArea.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.TextArea { /* compiled code */ }

public fun javafx.scene.layout.Pane.textfield(value: kotlin.String? /* = compiled code */, op: (javafx.scene.control.TextField.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.TextField { /* compiled code */ }

public fun javafx.scene.layout.Pane.titledPane(title: kotlin.String, node: javafx.scene.Node): javafx.scene.control.TitledPane { /* compiled code */ }

public fun javafx.scene.layout.Pane.toolbar(vararg nodes: javafx.scene.Node, op: (javafx.scene.control.ToolBar.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.control.ToolBar { /* compiled code */ }

public fun javafx.scene.layout.Pane.vbox(spacing: kotlin.Double? /* = compiled code */, children: kotlin.collections.Iterable<javafx.scene.Node>? /* = compiled code */, op: (javafx.scene.layout.VBox.() -> kotlin.Unit)? /* = compiled code */): javafx.scene.layout.VBox { /* compiled code */ }

```

**Comment by edvin** [comment]
I just released TornadoFX 1.2.3 with the new builders, so now you can write:

```kotlin
override val root = HBox()

init {
    with(root) {
        tabpane {
            tab("Tab 1") {
                // Content for tab 1 here
            }
            tab("Tab 2) // and so forth :)
        }
    }
}
```

If you find any missing builders or have more questions about how to do stuff, don't hesitate to ask @thomasnield. (But create new issues :)

Thanks for helping out again @hastebrot :)

**Comment by thomasnield** [comment]
Awesome thanks! I may file PR's if I find any missing builders, so you guys can focus on the more complicated stuff you're working on.
